web/src/search_suggestion.ts

```diff
@@ -449,30 +490,33 @@ function get_topic_suggestions(last, terms) {
     });
 }
 
-function get_term_subset_suggestions(terms) {
+function get_term_subset_suggestions(terms: NarrowTerm[]): Suggestion[] {
     // For stream:a topic:b search:c, suggest:
     //  stream:a topic:b
     //  stream:a
     if (terms.length < 1) {
         return [];
     }
 
-    let i;
-    const suggestions = [];
+    const suggestions: Suggestion[] = [];
 
-    for (i = terms.length - 1; i >= 1; i -= 1) {
+    for (let i = terms.length - 1; i >= 1; i -= 1) {
         const subset = terms.slice(0, i);
         suggestions.push(format_as_suggestion(subset));
     }
 
     return suggestions;
 }
 
-function get_special_filter_suggestions(last, terms, suggestions) {
-    const is_search_operand_negated = last.operator === "search" && last.operand[0] === "-";
+function get_special_filter_suggestions(
+    last: NarrowTerm,
+    terms: NarrowTerm[],
+    suggestions: Suggestion[],
+): Suggestion[] {
+    const is_search_operand_negated = last.operator === "search" && last.operand.startsWith("-");
     // Negating suggestions on is_search_operand_negated is required for
     // suggesting negated terms.
-    if (last.negated || is_search_operand_negated) {
+    if (is_search_operand_negated || last.negated) {
```

Comment:
Why did you swap these here?

Reply:
Honestly, this is a bit hacky. I am not sure if there is a better way to write this or not, but here `last.negated` can be `undefined`, which means we have to write a condition `last.negated ?? is_search_operand_negated`. Writing it like this is problematic because in this case, instead of performing a logical OR, the condition will just consider the value of `last.negated `(when it is not `undefined`), which will affect the logic.

Reply:
what about `if (last.negated === true || is_search_operand_negated) {`, to be more clear about what's being checked here?